### PR TITLE
Heatflow map

### PR DESCRIPTION
**Pull request by jaustermann**
The text output seems to work (i.e. the sum of all the outputted heatflow equals the top boundary heatflow in the heatflow statistics). The visualization output does not work yet. It compiles and runs but does not output the heat_flux_map as an option that can be selected in paraview. I ran the convection-box.prm cookbook as an example to test this and added the heat flux map to the postprocessor subsection. 

subsection Postprocess
  set List of postprocessors = dynamic topography, heat flux map, velocity statistics, temperature statistics, heat flux statistics, visualization

  subsection Visualization
    set List of output variables      = dynamic topography, heat flux map
    set Time between graphical output = 0.01
  end
end

### CODE REVIEW

**Comment by bangerth** [comment]
Your description seems to suggest that this is work in progress? Let us know if you want help!

(You'll probably also want to throw away the first few patches and only pick the last one.)

**Comment by jaustermann** [comment]
Yep, work in progress! I talked to Rene about this today and he couldn't find the issue right away so suggested sending the pull request so that it could be looked at through github. As described above this is a heat flow postprocessor that outputs the heatflow at each element on the top boundary as (1) a text file and (2) visual output for paraview or visit. Part (1) works I think but part (2) doesn't because even though it's specified to be outputted for visualization it doesn't come up in paraview. Sorry about the extra commits and the line change in main.cc ... I always forget the drill! I'll adjust that in the next update.

**Comment by gassmoeller** [comment]
Try deriving your visualization plugin from `CellDataVectorCreator<dim>` as well. Take a look at the first line of the class definition of artificial_viscosity.h (line 41) and just copy the structure. 

The problem was that we have two ways to create a visualization plugin: For cell data (like your plugin) you should derive from `CellDataVectorCreator<dim>` and implement the `execute` function as you did. For data that is available at quadrature points you can derive from `DataPostprocessorScalar<dim>` and `Interface<dim>` and then you need to implement the `compute_derived_quantities_vector` function.

**Comment by gassmoeller** [comment]
If that works please also add a test for each of the plugins (as small as possible, since you need to check in the output files as well).

File: source/postprocess/visualization/heat_flux_map.cc [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  {
+    namespace VisualizationPostprocessors
+    {
+
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(HeatFluxMap,
+                                                  "heat flux map",
+                                                  "A visualization output object that generates output for "
+                                                  "the heatflux across the top boundary. The heat flux is computed in "
+                                                  "outward direction, i.e., from the domain to the outside, "
+                                                  "using the formula $k \\nabla T \\cdot \\mathbf n$, where "
```
Comment (bangerth) [inline source/postprocess/visualization/heat_flux_map.cc]: should it not be the *negative* of this quantity?

File: source/postprocess/visualization/heat_flux_map.cc [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                                                  "$k$ is the thermal conductivity as reported by the material "
+                                                  "model, $T$ is the temperature, and $\\mathbf n$ is the outward "
+                                                  "normal. Note that the quantity so computed does not include "
+                                                  "any energy transported across the boundary by material "
+                                                  "transport in cases where $\\mathbf u \\cdot \\mathbf n \\neq 0$.")
```
Comment (bangerth) [inline source/postprocess/visualization/heat_flux_map.cc]: maybe add a sentence that the *integrated* heatflux for each part of the boundary can be obtained by the "heat flux" postprocessor; maybe also add a sentence there that the *pointwise* heat flux can be obtained using the "heat flux map" visualization postprocessor

Comment (jaustermann) [inline source/postprocess/visualization/heat_flux_map.cc]: Will do.

File: source/postprocess/visualization/heat_flux_map.cc (hunk 1 above)
Comment (jaustermann) [inline source/postprocess/visualization/heat_flux_map.cc]: I think that is right, in which case I will also correct that in the postprocess/heat_flux_map.cc and the heat flux statistics instantiations.

**Comment by jaustermann** [comment]
@gassmoeller Should I add a test for each plugin or is one that checks both enough?

**Pull request closed without merging**